src/python/pants/util/memo.py

```diff
@@ -4,8 +4,18 @@
 import functools
 import inspect
 from contextlib import contextmanager
+from typing import Any, Callable, Optional, TypeVar
 
-from pants.util.meta import classproperty, staticproperty
+from pants.util.meta import P, classproperty, staticproperty
+
+
+# NB: We liberally ignore MyPy issues in the implementations of these decorators. This is because
+#  it would be hard to both type the implementation correctly and get the call sites of the
+#  decorators properly typed. We care far more about the latter use case.
```

Comment:
I think this might be an implicit understanding -- not sure this comment is necessary, unless there are specific issues we want to put a TODO on?

Reply:
I'll remove it. I don't think it would ever be a good use of our time to try to get MyPy to understand the implementation. We have tests and enough usage to confirm it works.